File: LeetCode/1542. Find Longest Awesome Substring/solution.py
class Solution:
    def longestAwesome(self, s):
        res, cur, n = 0, 0, len(s)
        seen = [-1] + [n] * 1024
        for i, c in enumerate(s):
            cur ^= 1 << int(c)
            for a in range(10):
                res = max(res, i - seen[cur ^ (1 << a)])
            res = max(res, i - seen[cur])
            seen[cur] = min(seen[cur], i)
        return res        


# class Solution:
#     def longestAwesome(self, s: str) -> int:
#         cur, res = 0, 1
#         state = {0: -1}
#         for i, ch in enumerate(s):
#             ch = int(ch)
#             cur ^= (1<<ch)
#             if cur not in state:
#                 state[cur] = i 
#             else:
#                 res = max(res,i-state[cur])
#             for k in range(10):
#                 nxt = cur ^ (1<<k)
#                 if nxt in state:
#                     res = max(res,i-state[nxt])
#         return res

File: LeetCode/1542. Find Longest Awesome Substring/solution.cpp
class Solution {
public:
    int longestAwesome(string s) {
        int n = s.length();
        vector<int> dp(1<<10,n);
        dp[0] = -1;
        int res = 1, cur = 0;
        for (int i = 0; i < n; i++){
            int j = s[i]-'0';
            cur ^= (1<<j);
            int pi = dp[cur];
            for (int k = 0; k < 10; k++)
                pi = min(pi, dp[cur^(1<<k)]);
            res = max(res, i-pi);
            dp[cur] = min(dp[cur],i);
        }
        return res;
    }
};